### ISSUE TITLE
Code serialization builtins

### ISSUE THREAD

**Issue by pchiusano** [unisonweb/unison#1728]
**See [this comment](https://github.com/unisonweb/unison/issues/1728#issuecomment-712426285) for the latest design.**

The rest of this comment was the starting design and can be ignored unless you're just interested in following along with the design process that led to the above. 

---

Here's an API sketch for primitives for shipping code around at runtime. This would be used as a building block for network protocols that sync needed dependencies on the fly. This is a low-level, dynamically-typed API, not something you'd directly program against. These are the sort of primitives you'd use when providing an implementation of a `Remote` ability or some such:

```Haskell
builtin type Code

-- Convert any value to a `Code`, discarding all type information
Code.code : a -> Code
-- Is the local runtime missing a copy of the provided `Link`?
Code.isMissing : Link ->{IO} Boolean

-- Obtain the dependencies of a `Code`
Code.dependencies : Code -> [Link]

-- Serialization and deserialization
Code.serialize : Code -> Bytes
Code.deserialize : Bytes -> Either Text Code

-- Load a `Code` value - you better know what you're doing as this is effectively a cast
Code.load : Code ->{IO} (Optional a)
```

So any value can be converted to `Code`, the dependencies of a `Code` can be enumerated, and `Code` can be serialized and deserialized. A typical protocol might involve converting a value to `Code`, sending that over the network. The other side deserializes the `Code`, checks that it `Code.has` all the dependencies of that `Code`, and if yes, calls `Code.load`, and if no, contacts the sender (or perhaps a central `Code` cache) to retrieve any needed dependencies. 

With these protocol, the recipient will be assuming it's receiving a value of some particular type (like `'{Blah} ()`) and will be acting on that assumption.

**Comment by pchiusano** [comment]
@dolio I'm not totally clear on whether we even care about type dependencies, or if we just care about term dependencies. Like if Alice sends Bob a function `f` that references the `Foo` type, does Bob need a copy of `Foo`? Or does he just need a copy of the other functions that `f` references?

I think I might have answered my own question - the runtime will probably need to know at least the arities of all constructors / ability operations in order to generate the correct code.

**Comment by dolio** [comment]
So far I've been operating under the assumption that anything that would require knowing about constructor arities and whatnot would already be done before getting to the level that this `Code` type exists at.

However, I'm not certain that the various 'works for everything somehow' operations don't need to know about types.

What would cause `Code.load` to return `None`?

**Comment by pchiusano** [comment]
> What would cause `Code.load` to return `None`?

I was thinking it's when one of the dependencies of the code hasn't been loaded into the local cache. But that seems funky... like this API doesn't even have any way of loading anything into the cache. It's also kind of weird that the cache is this external thing. What if we instead make the cache explicit? (So the user can manage it themselves). Now I think you can get away without even mentioning `IO`:

```Haskell
builtin type Cache -- basically a Map Link Code

Cache.empty : Cache
Code.cache : Cache -> Link -> Code -> Cache
Code.load : Cache -> Code -> Optional a
```

A server might keep this in like an `MVar` or whatever.

> So far I've been operating under the assumption that anything that would require knowing about constructor arities and whatnot would already be done before getting to the level that this Code type exists at.

I see what you're saying about the code already having baked in the info needed for pattern matching and whatnot. I think it just depends what the wire protocol is... if the thing being sent is like serialized `MCode`, then you don't need the types any more. If it's basically a binary form of the original AST and the recipient has to do some more compilation of it, then you probably do need the types.

**Comment by dolio** [comment]
Possibly the `Cache` thing could work.

Another question is what `Code` is supposed to be. It doesn't make sense to assume that every value has a readily available 'source code' form that makes sense to turn into a definition that can be referenced. So this design seems to be conflating multiple notions.

>  if the thing being sent is like serialized `MCode`, then you don't need the types any more.

Not just `MCode`. There is a bunch of processing that happens before that stage, and at some point, all the necessary information about data types and such has been resolved, and code that doesn't need to look that up anymore has been generated. I thought this was an API for sending around such intermediate level code, not unison source code.

**Comment by pchiusano** [comment]
> I thought this was an API for sending around such intermediate level code, not unison source code.

That's right. It's not Unison source code. It's just some intermediate representation which runtime values can be converted to/from, that can be serialized, and that has a way of enumerating dependencies. (Maybe `Code` is a bad name and we should call it `IR` or something?).

> Not just MCode. There is a bunch of processing that happens before that stage, and at some point, all the necessary information about data types and such has been resolved, and code that doesn't need to look that up anymore has been generated.

That sounds nice! Basically, you can have `Code` correspond to one of those intermediate stages that has already baked in all the info about types that it needs. Now you don't need to even enumerate type dependencies or load them.

So, putting that all together, how about:

```Haskell
builtin type Code
builtin type Cache

-- Convert any value to a `Code`, discarding all type information
Code.code : a -> Code
-- Is the code cache missing the code corresponding to the provided `Link`?
Code.isMissing : Cache -> Link.Term -> Boolean

-- Obtain the dependencies of a `Code`
Code.dependencies : Code -> [Link.Term]

-- Serialization and deserialization
Code.serialize : Code -> Bytes
Code.deserialize : Bytes -> Either Text Code

Cache.empty : Cache
Cache.insert : Link -> Code -> Cache -> Cache
-- might want some other `Cache` functions too

-- Load a `Code` value - you better know what you're doing as this is effectively a cast
Code.load : Cache -> Code -> Optional a
```

**Comment by atacratic** [comment]
> It doesn't make sense to assume that every value has a readily available 'source code' form

Just elaborating on @dolio's point here:  is the argument to `Code.code` being used as a value or a term?  Or to put it another way, is `Code.code (2 + 2)` equal to `Code.code 4`?  If not, then you lose the `x = y implies f x = f y` property not just for `Code.code` but potentially any function than invokes it.  Basically `Code.code` has some metaprogramming quasiquoting stuff going on inside, but it's dressed up as if it's a regular function that conforms to a type signature.

**Comment by dolio** [comment]
I still think there needs to be a `Code` and a `Value` type that are separate. Values reference code, but aren't necessarily equivalent to code.

If the `Cache` is going to be explicit, too, it probably doesn't make sense for `code` to not take a `Cache` argument, either.  But then I'm not sure if it's really a good idea, because it means you could try to get a serializable value from a `Cache` that isn't capable of actually doing it. So, I guess, what is the advantage of making these explicit and first class values? The actual runtime still has an implicit analogue of these things, so it is misleading to think that you can (de)serialize something with an arbitrary `Cache` and have it make sense as a value with respect to that implicit runtime.

**Comment by atacratic** [comment]
> as if it's a regular function that conforms to a type signature

Maybe that bit was wide of the mark.  Maybe the problem is if `Code` supports equality comparison.  But then `Universal.==` is meant to cover everything...

**Comment by dolio** [comment]
@atacratic You don't have to worry about `2+2` producing a different thing than `4`, at least. This isn't like a quoting mechanism.

The problem is more like `f 2` where `f x y = x + y`. The runtime value will be a closure representation that references `f` and a representation of `2`. This closure representation is not the same thing as `g` where `g y = 2 + y` or something that makes sense to treat as a 'named' declaration. This particular example is simple, but it could be much more complex, and I don't think trying to design around making it happen would be very good.

**Comment by atacratic** [comment]
@dolio ah OK, maybe the penny is dropping, thank you.  

When would you have a value but not have all its dependencies?  Is it only after a `Code.load`?  

Other stuff:
- Is `Code.dependencies` returning the full transitive dependencies, or just one level?  Hopefully the latter as the former would be hard to scale well.
- if `Code.isMissing` returns true, what clues does the caller have about where to go find it?  Not seeing how you'll tie this into a code distribution system that's aware of multiple nodes.

**Comment by dolio** [comment]
The intention is to allow some amount of the protocol for migrating values to be programmable. So, if you did `deserialize` to get a value/code, you might not have all the dependencies, and need to negotiate with wherever you got the bytes from to get what you're missing. I guess in general you could use this to write protocols where you're told to contact a third party as well.

**Comment by pchiusano** [comment]
Simplifying after discussion with Dan. No first-class cache and separating `Code` and `Value`. Still TBD on handling insertion into the cache. We thought it would be nice to have it fail if the `Code` you are inserting has missing dependencies not found in the cache, but then realized that might not work when inserting elements that are part of a cycle. Going to do some more thinking on how to handle that.

```Haskell
builtin type Code
builtin type Value

-- Convert any value to a `Code`, discarding all type information
Code.value : a -> Value

-- Is the code cache missing the code corresponding to the provided `Link`?
Code.isMissing : Link.Term ->{IO} Boolean

-- Obtain the dependencies of a `Code`
Code.dependencies : Code -> [Link.Term]
Value.dependencies : Value -> [Link.Term]

-- Serialization and deserialization
Code.serialize : Code -> Bytes
Code.deserialize : Bytes -> Either Text Code
Value.serialize : Value -> Bytes
Value.deserialize : Bytes -> Either Text Value

Cache.insert_ : Link.Term -> Code ->{IO, Exception [Term.Link]} ()
-- This has no effect and returns a list of missing dependencies of the given `Code`
-- so you have to insert things into the cache in dependency order
Cache.insert : Link.Term -> Code ->{IO} [Term.Link]
Cache.lookup : Link.Term ->{IO} Optional Code

-- might want some other cache management functions but can sort out needed

-- Load a `Code` value - you better know what you're doing as this is effectively a cast
Value.load : Value ->{IO} Either [Term.Link] a
```

**Comment by pchiusano** [comment]
@dolio I think just have a batched cache insert:

```Haskell
-- For this to have an effect, all the combined dependencies must be in the 
-- cache already or included in the input list. Otherwise, the missing dependencies
-- are returned and this has no effect. 
Cache.inserts : [(Link.Term, Code)] ->{IO} [Term.Link]
```

And don't even bother with the single-element version (it can be defined in terms of the above).

I imagine you'd just collect up a big ol' list of all the missing dependencies and load them all in one go using the above function. This avoids you needing to do anything special to deal with cycles.

**Comment by pchiusano** [comment]
Okay, here's a final draft, synthesizing all of the above:

```Haskell
namespace runtime

builtin type Code
builtin type Value

-- Convert any value to a `Code`, discarding all type information
Value.value : a -> Value

-- Is the code cache missing the code corresponding to the provided `Link`?
Code.isMissing : Link.Term ->{IO} Boolean

-- Obtain the dependencies of a `Code`
Code.dependencies : Code -> [Link.Term]
Value.dependencies : Value -> [Link.Term]

-- Serialization and deserialization
Code.serialize : Code -> Bytes
Code.deserialize : Bytes -> Either Text Code
Value.serialize : Value -> Bytes
Value.deserialize : Bytes -> Either Text Value

-- For this to have an effect, all the combined dependencies must be in the 
-- cache already or included in the input list. Otherwise, the missing dependencies
-- are returned and this has no effect. 
Code.cache_ : [(Link.Term, Code)] ->{IO} [Term.Link]

Code.cache : [(Link.Term, Code)] ->{Exception [Term.Link], IO} [Term.Link]
Code.cache defs = match Code.cache_ defs with
  [] -> []
  missing -> Exception.raise missing

Code.lookup : Link.Term ->{IO} Optional Code
-- might want some other cache management functions eventually

-- Load a `Code` value - you better know what you're doing as this is effectively a cast
Value.load : Value ->{IO} Either [Term.Link] a

-- maybe Code.toValue : Code -> Value
```

**Comment by dolio** [comment]
I think that all looks sensible.

**Comment by ceedubs** [comment]
I think that this is resolved as of #1748.
